### PR TITLE
AgentView support

### PR DESCRIPTION
**Pull request by aaronschachter**
#### What's this PR do?
- Namespaces Mobile Commons specific config variables with a `MOBILECOMMONS_` prefix:
  - `MOBILECOMMONS_OIP_CAMPAIGNBOT`, `MOBILECOMMONS_OIP_DONORSCHOOSEBOT`, `MOBILECOMMONS_OIP_DONORSCHOOSEBOT_ENDED`, `MOBILECOMMONS_OIP_DONORSCHOOSEBOT_ERROR`
- Adds a new `GAMBIT_CMD_MEMBER_SUPPORT` config variable. If the member texts in the value during a CampaignBot conversation, they are opted into a Mobile Commons OIP defined by new config variable `MOBILECOMMONS_OIP_AGENTVIEW`.
  - The AgentView OIP conversation should be set to **NOT** trigger any mData's, which means all subsequent messages the member sends will be added to the Mobile Commons Agent View queue for member support to respond to, until they text in another Mobile Commons keyword.
  - Sends new CampaignBot `member_support` message to the member when they are opted into the Agent View OIP. This message should contain the `{{keyword_continue}}` variable to inform the member how to continue in CampaignBot after they have submitted their support questions.  `{{keyword_continue}}` renders our new `MOBILECOMMONS_KEYWORD_CAMPAIGNBOT`config variable.
  - The `MOBILECOMMONS_KEYWORD_CAMPAIGNBOT` should be set to the keyword defined on our CampaignBot mData, which posts to chatbot without a `campaign` (the same mData that our `MOBILECOMMONS_OIP_CAMPAIGNBOT` posts to). This allows the member to text in a keyword to continue where they left (loading their saved `current_campaign`).
  - The end result of all this makes our `member_support` copy look something like this: 
    
    > Text back your question and I'll do my best to answer back within 24 hours. If you'd like to continue on {{title}}, text back {{keyword_continue}}.
- Adds additional Gambit Jr. support for `{{cmd_reportback}}` and `{{cmd_member_support}}` for Gambit commands to use in CampaignBot messages.  Here's what an example `menu_completed` could read like: 
  
  > Got you down for {{quantity}} {{rb_noun}} {{rb_verb}}. If you have {{rb_verb}} more, text {{cmd_reportback}}. If you have a question, text {{cmd_member_support}}
- Refactors `isCommandCacheClear(req)` and `isCommandReportback(req)` to single `isCommand(req, type)` function with expected `type` values of `cache_clear`, `reportback`, and `member_support`
#### How should this be reviewed?

👀 
#### Relevant tickets
#606
#### Checklist
- [x] Documentation added for new features/changed endpoints.
- [x] Added required config vars on staging 
- [x] Tested on staging.
- [ ] Added required config vars on production

### CODE REVIEW

**Comment by aaronschachter** [comment]
Latest commits have been working well on staging, until Mobile Commons requests back to my phone display as failed. Don't see any errors in our Gambit app logs though 😕 

**Note** I had to alter our chatbot Mobile Commons OIP to force sending MMS today in order to avoid rendered message content truncating at 160 chars -- wonder if that could be the issue?

screen shot 2016-10-06 at 9 57 21 pm

**Comment by aaronschachter** [comment]
Following up on the last comment, failed messages seem related to my profile specifically and due to a Mobile Commons "aggregator problem" with my phone and MMS messages (not SMS messages).

I did change our CampaignBot Staging OIP to "Force MMS" for the time being while we're in development -- this change was made to avoid our Gambit chatbot copy getting truncated (and put off on engineering for non-MMS members until we have our prototype finished).

File: api/controllers/CampaignBotController.js [hunk 1]
```diff
@@ -36,7 +31,7 @@ class CampaignBotController {
   collectReportbackProperty(req, property, ask) {
     this.debug(req, `collectReportbackProperty:${property}`);
 
-    if (req.query.campaign || ask) {
+    if (req.query.campaign || ask || this.isKeywordCampaignBot(req)) {
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]: Why this word order? Wouldn't `isCampaignBotKeyword()` read better?

File: api/controllers/CampaignBotController.js [hunk 2]
```diff
@@ -239,39 +234,51 @@ class CampaignBotController {
...
+    const configValue = process.env[configName];
+    if (!configValue) {
+      logger.error(`${this.loggerPrefix(req)} process.env.${configName} is undefined`);
 
       return false;
     }
 
-    const input = helpers.getFirstWord(req.incoming_message);
-    const isCommand = input && input.toUpperCase() === cmdClear.toUpperCase();
+    const isCommand = this.parseCommand(req) === configValue.toUpperCase();
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]: It might be a bad idea to name the variable the same as the function. Kinda confusing.
In my opinion, simple `result` would suffice.

File: api/controllers/CampaignBotController.js [hunk 3]
```diff
@@ -239,39 +234,51 @@ class CampaignBotController {
...
+    if (!type) {
+      return false;
+    }
 
-    if (req.user.role !== 'staff' && req.user.role !== 'admin') {
-      logger.warn(`${this.loggerPrefix(req)} unauthorized clear cache`);
+    const configName = `GAMBIT_CMD_${type.toUpperCase()}`;
+    const configValue = process.env[configName];
+    if (!configValue) {
+      logger.error(`${this.loggerPrefix(req)} process.env.${configName} is undefined`);
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]: Why did you change it from warn to error?

File: api/controllers/CampaignBotController.js [hunk 4]
```diff
@@ -239,39 +234,51 @@ class CampaignBotController {
...
 
       return false;
     }
 
-    const input = helpers.getFirstWord(req.incoming_message);
-    const isCommand = input && input.toUpperCase() === cmdClear.toUpperCase();
+    const isCommand = this.parseCommand(req) === configValue.toUpperCase();
+
+    if (type === 'clear_cache') {
+      const authorized = req.user.role && (req.user.role === 'staff' || req.user.role === 'admin');
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]: There should be a function for that.

File: api/controllers/CampaignBotController.js [hunk 5]
```diff
@@ -239,39 +234,51 @@ class CampaignBotController {
...
-   * @param {object} req - Express request
+   * Returns whether incoming request is the CampaignBot Mobile Commons keyword.
+   * @param {object} req
    * @return {bool}
    */
-  isCommandReportback(req) {
-    const firstWord = helpers.getFirstWord(req.incoming_message);
-    if (!firstWord) return false;
+  isKeywordCampaignBot(req) {
+    if (req.body.keyword) {
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]:
```js
if (!req.body.keyword) {
  return false;
}

// Continue as normal
```

File: api/controllers/CampaignBotController.js [hunk 6]
```diff
@@ -433,6 +447,9 @@ class CampaignBotController {
     msg = msg.replace(/{{rb_noun}}/gi, campaign.reportbackInfo.noun);
     msg = msg.replace(/{{rb_verb}}/gi, campaign.reportbackInfo.verb);
     msg = msg.replace(/{{rb_confirmed}}/gi, campaign.reportbackInfo.confirmationMessage);
+    msg = msg.replace(/{{cmd_reportback}}/gi, process.env.GAMBIT_CMD_REPORTBACK);
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]: Is there going to be more than one match? If no, `/g` is an overhead.

File: config/router-chatbot.js [hunk 7]
```diff
@@ -34,10 +37,19 @@ router.post('/', (req, res) => {
     return app.locals.controllers.donorsChooseBot.chatbot(req, res);
   }
 
-  const mobilecommonsOip = process.env.CAMPAIGNBOT_MOBILECOMMONS_OIP;
-  if (!mobilecommonsOip) {
-    logger.error('CAMPAIGNBOT_MOBILECOMMONS_OIP undefined');
+  let configured = true;
+  // Check for required config variables.
+  const settings = ['MOBILECOMMONS_OIP_CAMPAIGNBOT', 'MOBILECOMMONS_OIP_AGENTVIEW'];
+  settings.push('MOBILECOMMONS_KEYWORD_CAMPAIGNBOT');
```
Comment (sergii-tkachenko) [inline config/router-chatbot.js]: Maybe just add it to `concat` below?

Comment (sergii-tkachenko) [inline config/router-chatbot.js]: Or even just add all then to the array above.

File: config/router-chatbot.js [hunk 8]
```diff
@@ -34,10 +37,19 @@ router.post('/', (req, res) => {
...
+  settings.concat(['GAMBIT_CMD_MEMBER_SUPPORT', 'GAMBIT_CMD_REPORTBACK']);
+  settings.forEach((configVar) => {
+    if (!process.env[configVar]) {
+      logger.error(`undefined process.env.${configVar}`);
+      configured = false;
```
Comment (sergii-tkachenko) [inline config/router-chatbot.js]: Did you mean to end the loop when the first var is not found?

File: config/router-chatbot.js [hunk 9]
```diff
@@ -132,12 +148,18 @@ router.post('/', (req, res) => {
...
+        oip = process.env.MOBILECOMMONS_OIP_AGENTVIEW;
+      }
+
+      mobilecommons.chatbot(req.body, oip, msg);
 
       return res.send({ message: msg });
     })
     .catch(err => {
       controller.error(req, res, err);
+
```
Comment (sergii-tkachenko) [inline config/router-chatbot.js]: Extra line?

File: lib/helpers.js [hunk 10]
```diff
@@ -133,6 +133,18 @@ function getFirstWord(message) {
 };
 module.exports.getFirstWord = getFirstWord;
 
+
+/**
+ * Returns the uppercase first word of given message.
+ */
+module.exports.getFirstWordUppercase = function (message) {
+  const isText = message || getFirstWord(message);
```
Comment (sergii-tkachenko) [inline lib/helpers.js]: You shouldn't call `getFirstWord(message)` twice.

**Review by sergiitk — CHANGES_REQUESTED** [review]

File: api/controllers/CampaignBotController.js (hunk 1 above)
Comment (aaronschachter) [inline api/controllers/CampaignBotController.js]: I think this was forward thinking to an issue about to create about making use of the `body.keyword` property we get from a Mobile Commons request, where we'll have different keyword types. Signup, Continue. CampaignBot keyword here should really be continue.

I'm thinking this new function could work like the `isCommand(req, type)` does -- `isKeyword(req, type)` where type can be `'signup'` or `'continue'`.

File: config/router-chatbot.js (hunk 8 above)
Comment (aaronschachter) [inline config/router-chatbot.js]: I did write it that way at first -- but thought it could be useful to log each missing variable too.

File: config/router-chatbot.js (hunk 7 above)
Comment (aaronschachter) [inline config/router-chatbot.js]: Yeah this isn't as clean. I was debating on adding them all to 1 line adding using the `disable-eslint` -- would that be best @sergii-tkachenko ?

File: api/controllers/CampaignBotController.js (hunk 6 above)
Comment (aaronschachter) [inline api/controllers/CampaignBotController.js]: Good call, there shouldn't be.

File: config/router-chatbot.js (hunk 9 above)
Comment (aaronschachter) [inline config/router-chatbot.js]: I was adding the extra line in attempt to make the `return` more readable (like above on L 158). Per http://eslint.org/docs/rules/newline-before-return -- there's no "hard and fast rule". I'm fine with removing, but maybe it could make sense for us to add one (a rule?)

Comment (sergii-tkachenko) [inline config/router-chatbot.js]: Oh, good. I missed (forgot?) this rule. Thank you for pointing it out, will be doing the same from now on!

File: api/controllers/CampaignBotController.js [hunk 11]
```diff
@@ -239,39 +234,46 @@ class CampaignBotController {
...
+      logger.warn(`${this.loggerPrefix(req)} process.env.${configName} is undefined`);
 
       return false;
     }
 
-    const input = helpers.getFirstWord(req.incoming_message);
-    const isCommand = input && input.toUpperCase() === cmdClear.toUpperCase();
+    const result = this.parseCommand(req) === configValue.toUpperCase();
+
+    if (type === 'clear_cache') {
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]: Do you want to move this line condition above the `const result` line?
It looks like it doesn't rely on anything so we don't need to calculate result when the type is 'clear_cache', right?

File: api/controllers/CampaignBotController.js [hunk 12]
```diff
@@ -442,7 +454,8 @@ class CampaignBotController {
       msg = msg.replace(/{{quantity}}/gi, quantity);
     }
 
-    if (req.query.campaign && req.signup && req.signup.draft_reportback_submission) {
+    const continuing = req.body.keyword && req.signup && req.signup.draft_reportback_submission;
```
Comment (sergii-tkachenko) [inline api/controllers/CampaignBotController.js]: Did you change it from `revisiting`? I thought it was more explicit.

**Review by sergiitk — CHANGES_REQUESTED** [review]

File: api/controllers/CampaignBotController.js (hunk 12 above)
Comment (aaronschachter) [inline api/controllers/CampaignBotController.js]: Ah ok i'll back to `revisiting` -- good point

File: api/controllers/CampaignBotController.js (hunk 11 above)
Comment (aaronschachter) [inline api/controllers/CampaignBotController.js]: Great catch @sergii-tkachenko -- this was firing when it wasn't supposed to because we do need that `result` -- https://github.com/DoSomething/gambit/pull/654/commits/41cfc05d239832cd6f04ae50379454b394ae07ba

**Comment by aaronschachter** [comment]
@sergii-tkachenko 👍  ?

**Review by sergiitk — APPROVED** [review]
👍 👍 👍
